File: Snooper-Spring/src/main/java/org/springframework/http/client/SpringHttpRequestTransformer.java
package org.springframework.http.client;

import com.prateekj.snooper.model.HttpCall;

import org.apache.commons.io.IOUtils;
import org.springframework.http.HttpRequest;

import java.io.IOException;
import java.nio.charset.Charset;

public class SpringHttpRequestTransformer {

  public HttpCall transform(HttpRequest httpRequest, byte[] requestPayload, ClientHttpResponse httpResponse) throws IOException {
    return new HttpCall.Builder()
      .withUrl(httpRequest.getURI().toString())
      .withPayload(new String(requestPayload))
      .withMethod(httpRequest.getMethod().toString())
      .withResponseBody(IOUtils.toString(httpResponse.getBody(), Charset.defaultCharset()))
      .withStatusCode(httpResponse.getRawStatusCode())
      .withStatusText(httpResponse.getStatusCode().getReasonPhrase())
      .withRequestHeaders(httpRequest.getHeaders())
      .withResponseHeaders(httpResponse.getHeaders())
      .build();
  }
}

File: Snooper-Spring/src/main/java/org/springframework/http/client/SpringHttpRequestTransformer.kt
package org.springframework.http.client

import com.prateekj.snooper.networksnooper.model.HttpCall
import org.springframework.http.HttpRequest
import java.io.IOException

class SpringHttpRequestTransformer {

  @Throws(IOException::class)
  fun transform(
    httpRequest: HttpRequest,
    requestPayload: ByteArray,
    httpResponse: ClientHttpResponse
  ): HttpCall {
    return HttpCall.Builder()
      .withUrl(httpRequest.uri.toString())
      .withPayload(String(requestPayload))
      .withMethod(httpRequest.method.toString())
      .withResponseBody(httpResponse.body.bufferedReader().use { it.readText() })
      .withStatusCode(httpResponse.rawStatusCode)
      .withStatusText(httpResponse.statusCode.reasonPhrase)
      .withRequestHeaders(httpRequest.headers)
      .withResponseHeaders(httpResponse.headers)
      .build()
  }

  @Throws(IOException::class)
  fun transform(httpRequest: HttpRequest, requestPayload: ByteArray, e: Exception): HttpCall {
    return HttpCall.Builder()
      .withUrl(httpRequest.uri.toString())
      .withPayload(String(requestPayload))
      .withMethod(httpRequest.method.toString())
      .withRequestHeaders(httpRequest.headers)
      .withError(e.toString())
      .build()
  }
}
